File: frontend/src/types/feedback.ts
/**
 * Feedback-related TypeScript types
 * CRITICAL: Feedback data feeds Phase 2 memory/learning system
 * FIXED: Aligned with backend FeedbackStatsResponse schema
 */

export interface Feedback {
  id: number;
  suggestion_id: number;
  sample_id: number;
  action: FeedbackAction;
  final_label: number;
  iteration: number;
  review_time_seconds?: number;
  created_at: string;
}

export interface FeedbackWithDetails extends Feedback {
  current_label?: number;
  suggested_label?: number;
  original_label?: number;
  confidence_score?: number;
  detection_info?: {
    confidence_score: number;
    anomaly_score: number;
    priority_score: number;
  };
}


export interface FeedbackStats {
  dataset_id: number;
  total_feedback: number;
  accepted: number;     
  rejected: number;     
  modified: number;     
  acceptance_rate: number;
  avg_review_time?: number;
}

export interface PatternRangeData {
  total: number;
  accepted:number;
  acceptance_rate: number;
}
export interface FeedbackPatterns {
  dataset_id: number;
  iteration?: number;
  patterns_found: number;
  acceptance_by_confidence?: {
    [range: string]: PatternRangeData; 
  };
  acceptance_by_priority?: {
    [priority: string]: PatternRangeData; 
  };
  insights?: string[];
  message?: string; 
}

export interface FeedbackListResponse {
  feedback: Feedback[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export type FeedbackAction = 'accept' | 'reject' | 'modify';

export interface FeedbackFilters {
  dataset_id?: number;
  iteration?: number;
  action?: FeedbackAction;
  page?: number;
  page_size?: number;
}

File: backend/schemas/feedback.py
"""
Feedback schemas for API request/response validation
"""
from pydantic import BaseModel, Field, ConfigDict
from datetime import datetime
from typing import Optional, List


class FeedbackBase(BaseModel):
    """Base feedback schema"""
    suggestion_id: int
    sample_id: int
    action: str = Field(..., pattern="^(accept|reject|modify)$")
    final_label: int
    iteration: int
    
    model_config = ConfigDict(protected_namespaces=())


class FeedbackCreate(FeedbackBase):
    """Schema for creating feedback"""
    review_time_seconds: Optional[float] = None
    
    model_config = ConfigDict(protected_namespaces=())


class FeedbackResponse(BaseModel):
    """Schema for feedback response"""
    id: int
    suggestion_id: int
    sample_id: int
    action: str
    final_label: int
    iteration: int
    review_time_seconds: Optional[float] = None
    created_at: datetime
    
    model_config = ConfigDict(from_attributes=True, protected_namespaces=())


class FeedbackWithDetails(FeedbackResponse):
    """Feedback with additional context"""
    current_label: Optional[int] = None
    suggested_label: Optional[int] = None
    original_label: Optional[int] = None
    confidence_score: Optional[float] = None
    
    model_config = ConfigDict(from_attributes=True, protected_namespaces=())


class FeedbackStatsResponse(BaseModel):
    """Statistics about feedback for a dataset"""
    dataset_id: int
    total_feedback: int
    accept_count: int
    reject_count: int
    modify_count: int
    acceptance_rate: float
    avg_review_time: Optional[float] = None
    
    model_config = ConfigDict(protected_namespaces=())


class FeedbackPatternResponse(BaseModel):
    """Pattern analysis from feedback"""
    dataset_id: int
    iteration: int
    most_accepted_class: Optional[int] = None
    most_rejected_class: Optional[int] = None
    high_confidence_acceptance_rate: float
    low_confidence_acceptance_rate: float
    
    model_config = ConfigDict(protected_namespaces=())


class FeedbackListResponse(BaseModel):
    """Response for listing feedback"""
    feedback: List[FeedbackResponse]
    total: int
    page: int
    page_size: int
    total_pages: int
    
    model_config = ConfigDict(protected_namespaces=())